fix: update case for alpine:edge correct vuln feed

spiffcs opened the pull request:
Partial Fix for #964 

## Summary
When scanning `alpine:edge` with grype the incorrect vulnerability namespace would be used causing false positives that would already be fixed in the upstream source:

https://security.alpinelinux.org/branch/edge-main
https://gitlab.alpinelinux.org/ariadne/secfixes-tracker

This fix treats nil distro version segments (a product of a version string that could not be parsed), when distro is of type `alpine`, as the signal for returning the `alpine:edge` namespace.

### Notes

Please read the related issue for upstream recommendations on reducing the false positives produced by this PR. 

Signed-off-by: Christopher Phillips <christopher.phillips@anchore.com>

wagoodman commented on grype/db/v5/namespace/index.go:
```diff
@@ -118,6 +118,15 @@ func (i *Index) NamespacesForDistro(d *grypeDistro.Distro) []*distro.Namespace {
 
+	// Fall back to alpine:edge if no version segments found
+	// alpine:edge is labeled as alpine-x.x_alphaYYYYMMDD
+	if versionSegments == nil && d.Type.String() == "alpine" {
```
I think there is a `grypeDistro.Alpine` to compare against instead of the string (or does that not work in this case?)

luhring replied:
Re: checking the version, my suggestion here would be to find a way to test this value to see if it matches the `x.y.z` format mentioned in [this comment](https://github.com/anchore/grype/issues/964#issuecomment-1290888755). The logic right now is still just a bit off, because there can be cases where the version is something like `1.2.3-beta`, so `versionSegments` would not be `nil`, and this code won't use edge when it actually should use edge.

spiffcs replied:
Updated and added the case for the distro to match the version against a valid regex for `x.y.z` falling back to edge if `x.y.z` is unmatched.

wagoodman commented on grype/db/v5/namespace/index.go:
```diff
@@ -118,6 +118,15 @@ func (i *Index) NamespacesForDistro(d *grypeDistro.Distro) []*distro.Namespace {
+	// Fall back to alpine:edge if no version segments found
+	// alpine:edge is labeled as alpine-x.x_alphaYYYYMMDD
+	if versionSegments == nil && d.Type.String() == "alpine" {
+		distroKey := fmt.Sprintf("%s:%s", strings.ToLower(d.Type.String()), "edge")
```
should we check to see if the raw version has something like `alpha` or `beta`? (from https://github.com/anchore/grype/blob/main/grype/distro/distro.go#L16)

spiffcs replied:
Updated this with the new test cases to match against `x.y.z` and then falling back to edge so no need to check for `alpha` or `beta` in this case

wagoodman commented on grype/db/v5/namespace/index_test.go:
```diff
@@ -137,124 +138,155 @@ func TestIndex_NamespacesForDistro(t *testing.T) {
 		},
 		{
+			name:   "alpine edge version matches edge namespace",
+			distro: &osDistro.Distro{Type: osDistro.Alpine, Version: nil, RawVersion: "3.17_alpha20221002", IDLike: []string{"alpine"}},
```
we should also add a case that covers the possibility if they start changing their versions to be semver compliant (or close) as an extra precaution (e.g. `3.17-alpha20221002` or `3.17.0-alpha20221002`). I realize this is unlikely, but knowing the namespace logic is resilient in cases like these is useful since the data can change out from underneath us (and the alternative is a code change when that happens).

spiffcs replied:
Test cases were added for this possibility to check for namespace resilience.

luhring commented:
Cross-posting this update for visibility 😃  https://github.com/anchore/grype/issues/964#issuecomment-1290888755

spiffcs commented:
@luhring and @kzantow this one has been updated. Nice find @luhring on the possibility of `versionSegments` - I've updated the logic and test cases to account for this.

luhring commented on grype/db/v5/namespace/index.go:
```diff
@@ -12,6 +13,8 @@ import (
 	syftPkg "github.com/anchore/syft/syft/pkg"
 )
 
+var alpineVersionRegularExpression = regexp.MustCompile(`^(\d+)\.(\d+)(?:\.(\d+))?(?:-r(\d+))?$`)
```
I think this regex should be changed to `^(\d+)\.(\d+)\.(\d+)$`. As written the regex will match version strings beyond the `x.y.z` rule, such as `3.16` and `1.2.3-r4`.

spiffcs replied:
^ Ahh yea - I thought that `x.y.z` would also include `x.y` as a derivative. I'll update this to be more strict and add a case for the r4

luhring commented on grype/db/v5/namespace/index_test.go:
```diff
@@ -137,124 +138,179 @@ func TestIndex_NamespacesForDistro(t *testing.T) {
+			name:   "alpine minor version matches minor version namespace",
 			distro: newDistro(t, osDistro.Alpine, "3.16", []string{}),
 			namespaces: []*distro.Namespace{
 				distro.NewNamespace("alpine", osDistro.Alpine, "3.16"),
```
This test case should expect the "edge" namespace, not "3.16"

luhring commented on grype/db/v5/namespace/index.go:
```diff
@@ -12,6 +13,8 @@ import (
 )
 
+var alpineVersionRegularExpression = regexp.MustCompile(`^(\d+)\.(\d+)(?:\.(\d+))?(?:-r(\d+))?$`)
+
```
For **lines 100-103** _(GitHub isn't letting me comment on those lines for some reason, sorry!)_:

If the Alpine version string is something like `3.16.0-r4`, we'll fall into the conditional case in these lines and potentially return the wrong namespace. The version string `3.16.0-r4` should map to `alpine:edge`, but instead the code will see that `len(versionSegments) == 3` and that the `alpine:3.16` namespace is present in the index, so I believe it will incorrectly return `alpine:3.16` in that scenario.

luhring replied:
It might be useful to add a test case to catch this scenario

luhring reviewed (commented):
This is getting close! Added some comments around a theme - attempting to clarify that the `x.y.z` rule means literally 3 digits, and anything else should map to the edge data. Hope this is helpful 😃

spiffcs commented:
Thanks for the follow-up @luhring and the catch about the minor vs patch namespace selection:
- I added a case for `rc4`
- I updated `alpine` to be a priority and checked first with the regex you suggested so that we don't fall into the more general logic of the other distros

luhring commented:
Awesome! This behavior looks correct to me 👍
